Reports ported: VAT Listing and VAT Intra

andreparames opened the pull request:
This is a straight migration of two reports to version 10.

andreparames committed:
Reports ported: VAT Listing and VAT Intra

sbidoul commented on l10n_be_vat_reports/wizard/l10n_be_partner_vat_listing.py:
```diff
@@ -0,0 +1,371 @@
+                      'tax_tag_03', 'tax_tag_45', 'tax_tag_49')
+                  AND l.partner_id IN %s
+                  AND l.date BETWEEN %s AND %s
+                  GROUP BY l.partner_id, p.name, p.vat) AS sub1
```
@andreparames the turnover computation is incorrect. 

In v9+, we must sum move lines where "tax_ids has one of (00,01,02,03,45,49)" to find the tax bases.

andreparames replied:
But that's what I do (in lines 78-83), it's just that I search using the XML-IDs. I find all the account_tax with a tag that has on of the following XML-IDs: 'tax_tag_00', 'tax_tag_01', 'tax_tag_02', 'tax_tag_03', 'tax_tag_45', 'tax_tag_49', which are defined in the v9 of l10n_be: https://github.com/odoo/odoo/blob/9.0/addons/l10n_be/account_tax_template.xml#L5

Is this incorrect? Should I be hardcoding the account_tax ids?

andreparames committed:
[FIX] Correctly calculate turnover and vat_amount

This commit fixes two problems with the calculation:

- The turnover tax was being fetched from the wrong field
- Move lines with taxes that had more than one tag where being
summed more than once
Use -balance for turnover and tax

andreparames commented:
@sbidoul thanks for the heads up! I've pushed a fix for that bug and another which I discovered during testing.

sbidoul commented:
As discussed:
- use  `-balance` for turnover and tax
- replace partition by some sort of where exists (btw, the list of tax tags ids could be prefetched with the ORM)
- there is probably a similar change to do for the vat_intra report
- a couple of line too long errors

sbidoul commented:
Or even better, `search()` for the taxes with the corresponding tax and then use `tax_line_id in %s` and `where exists` on the tax_ids relation.

andreparames committed:
[FIX] vat reports: Create XML records in current module

They were being created in l10n_be

andreparames committed:
vat reports: add test

andreparames commented:
I wrote them as CTEs, just tell me if you prefer the `search()` solution anyway.

I've also added a test for the VAT listing that checks one value in the resulting XML.

vdewulf commented:
Hello,
I tested locally the Annual Listing of VAT-Customer on version 9 (I just replaced the name of the file "__manifest__" by "__openerp__".
When there is no VAT number, it doesn't appear in the list, which I suppose is the intended behavior).

# Tested OK
One customer has 2 invoices. Each invoice is about 100€ hvat + 21% vat.
=> it doesn't appear on the list. OK
I created a 3th invoice (100€ hvat). When generating the Annual Listing wizard, there is a line with 300€ and 63€ vat and the corresponding VAT number.
=> OK

I also tested the limit amount by changing its value. The corresponding listing is correctly updated.

# To confirm
When there is no tax set on the invoice line, the corresponding amount does not appear in the total of the Base Amount.
=> Is it intended? (I'm thinking about the "régime de la franchise de TVA" where the company has to give such listing even if no VAT was invoiced to the customer). Should be confirmed by an accountant.

I tested the XML creation. The XML is created but I don't know the requirements to go futher on this testing.

# Tested Not OK
I tested the printing of the listing to a PDF. I got an error (AttributeError: 'list' object has no attribute 'ids').
Details of the log in attachment.
[Error_Printing_VAT_Listing_to_PDF.txt](https://github.com/OCA/l10n-belgium/files/1866489/Error_Printing_VAT_Listing_to_PDF.txt)

andreparames committed:
vat reports: improve backward compatibility

andreparames commented:
@vdewulf thanks for the test. The behavior in v10 was made more robust in Odoo, but I've pushed a change that should make it work on both versions. Can you try on v9 again?

vdewulf commented:
@andreparames I tested with the last version on v9 as proposed.
## Tested OK ##
I tested the printing of the listing to a PDF ant it worked this time.
I re-tested changing the value in Limit Amount and the result was correctly updated.
The refund invoices are well taken into account in the total amount in the list.
I compared the result with a listing made my an accoutant for a customer and the result is the same (except some rounding differences).

## Confirmed (cf. my previous comment) ##
For my question about the special "régime de franchise TVA", on the invoice, the tax 0% should be used so that these amount would appear in the listing (I suppose it's the case, I did not tested this specific case).

## Tested Not OK ##
I tried to generate the XML but a pop up appeared with the message "No email address associated with the company.".
However, I have an email set in the general settings of the company.

In the logs, following messages appeared when I clicked on the button to create the XML file:
`2018-04-11` 13:23:45,620 11969 WARNING medor-vat-listing openerp.addons.base.ir.ir_sequence: ir_sequence.get() and ir_sequence.get_id() are deprecated. Please use ir_sequence.next_by_code() or ir_sequence.next_by_id().
`2018-04-11` 13:23:45,635 11969 WARNING medor-vat-listing openerp.addons.base.ir.ir_sequence: ir_sequence.get() and ir_sequence.get_id() are deprecated. Please use ir_sequence.next_by_code() or ir_sequence.next_by_id().

Maybe it's due to my DB, bad configuration... ? I didn't test on another DB.

andreparames commented:
@vdewulf the email address must be on the Invoicing address of the company. I'm not sure if this still makes sense, but it was the original behavior. If you add one, it should work.

I'll fix those warnings, thanks for pointing them out!

vdewulf commented:
@andreparames Indeed, it worked. I had a message about the phone that was missing, I put one on the invoicing adress and then the XML could be generated.

In the XML, there are encoding issues (for example for the city of Liège: <City>LiÃ¨ge</City>)
I don't know if this should be fixed or not (does the administration accept this sort of file?).
For information I only tested the Annual Listing of VAT-Customer, not the other report.

Another improvement: If MIS Builder is installed together with the Belgian Templates (Belgium P&L, Balance Sheet, VAT Declaration), there is in the menu "Report" a sub-menu named "Belgium Statements" under which there is "VAT Declaration".
With this module l10n_be_vat_reports installed, there is another sub-menu named "Belfium Statements" under which there are the VAT Listing and the Intra Comm report.
Would it be possible to merge the submenus in one only?
See screenshot attached:
[Menu_Belgium_Statements.pdf](https://github.com/OCA/l10n-belgium/files/1899104/Menu_Belgium_Statements.pdf)

Thanks a lot!

sbidoul commented on l10n_be_vat_reports/wizard/l10n_be_partner_vat_listing.xml:
```diff
@@ -0,0 +1,118 @@
+    <menuitem
+        id="menu_finance_belgian_statement"
+        name="Belgium Statements"
+        parent="account.menu_finance_reports" groups="account.group_account_manager"/>
```
This menu already exists in l10n_be indeed.

See example use here: https://github.com/OCA/l10n-belgium/blob/10.0/l10n_be_mis_reports/data/mis_report_vat.xml#L12

andreparames replied:
Thanks for the warning!

sbidoul commented:
> In the XML, there are encoding issues (for example for the city of Liège: LiÃ¨ge)

@vdewulf Can you try opening the xml file in firefox? Or attach here so we can have a look.

vdewulf commented:
@sbidoul Here is a screenshot after opening in Firefox. As these are real date from our customer, I prefer not to put the whole file here.
![xml_open_in_firefox](https://user-images.githubusercontent.com/27895270/38623430-ee8b7caa-3da5-11e8-9979-5654a806503d.png)

sbidoul commented on l10n_be_vat_reports/wizard/l10n_be_partner_vat_listing.py:
```diff
@@ -0,0 +1,388 @@
+""" % annual_listing_data
+
+        data_file += data_begin + data_comp + data_client_info + data_end
+        file_save = base64.encodestring(data_file.encode('utf8'))
```
@andreparames here encode('ISO-8859-1') instead of utf8 as that's what indicated in the XML declaration above.

andreparames replied:
While this is clearly problematic, it was already outputting in UTF-8, so isn't it risky to change now? Should we change the declaration instead?

sbidoul replied:
Ok, let's try with changing the declaration

andreparames committed:
vat reports: upgrade sequence code

sbidoul commented on l10n_be_vat_reports/wizard/l10n_be_vat_intra.py:
```diff
@@ -0,0 +1,376 @@
+            '\n\t\t<ns2:Comment>%(comments)s</ns2:Comment>\n\t'
+            '</ns2:IntraListing>\n</ns2:IntraConsignment>') % (xml_data)
+        context = dict(self.env.context)
+        context['file_save'] = data_file
```
encode ISO-8859-1 too here?

andreparames committed:
vat reports: fixing menu references

andreparames commented:
Ok, I've merged the two menus and have updated the XML declarations.

andreparames committed:
vat reports: fix UI problems in VAT intra report

andreparames committed:
vat reports: improved intra report

sbidoul commented on l10n_be_vat_reports/wizard/l10n_be_vat_intra.py:
```diff
@@ -116,7 +109,7 @@ def _get_datas(self):
             raise UserError(_('Period code is not valid.'))
 
-        if not wiz_data.date_start > wiz_data.date_end:
+        if not wiz_data.date_start <= wiz_data.date_end:
```
remove the `not`

andreparames committed:
vat reports: fix encoding & remove pointless code

andreparames commented:
Thanks for the review. I've pushed some fixes to the Intra report, plus a test file.

vdewulf commented:
@andreparames Hi! Do you want me to test the last version? Just tell me! Thanks.

andreparames commented:
@vdewulf sure, the latest version should fix the two problems you identified (file encoding and duplicated menu).

vdewulf commented:
@andreparames 
### VAT listing
I tested OK the encoding fix and the unified menu. Seems OK for the VAT listing.
### Intracomm report
I tested the intracomm report. When I click on "preview" button, it raised an error :
`File "/home/virg/odoo-dev/odoo/addons/report/models/report.py", line 307, in get_action
    return Report.get_action(self._model, self._cr, self._uid, records.ids,`
`AttributeError: 'list' object has no attribute 'ids'`
while an XML is being created.
In the XML file, there is a False value in this code: <ns2:Code>False</ns2:Code>
Is it intended? What should be there?
![intracomm](https://user-images.githubusercontent.com/27895270/38943682-4eed8336-4332-11e8-91d7-fd3e9933b81e.png)

andreparames committed:
vat reports: small corrections

andreparames commented:
@vdewulf thanks for the review, I pushed an update that should fix the remaining problems.

vdewulf commented:
I came back to this PR. sorry for the delay!

VAT Listing
I retested the basic scenario, testing when there is no VAT number on a partner (it does not appaear in the list), when the limit is higher or lower than 250€ (the corresponding info are displayed).
I tested the printing, it worked fine.
I tested the XML generation, it worked fine.

Intracomm report
I tested the printing and the XML generation.
The False value detected previously has disapeared (I got S code instead).

I think that works fine, but I would like to double check with more data and and expert in accounting.

Okay for me, thanks for the work!

sbidoul commented:
Ok. Before merging we'll need to do some rework of the commits, to have a first commit that is exactly identical to the Odoo 8 version, and then add our changes on top, so copyright can be properly attributed to the part that comes from Odoo SA.

vdewulf commented:
@sbidoul Hello Stephane. Happy New Year 2019!
I'd like to know if this PR will soon be merged or not?
It will be need by our clients in march 2019 in v9.
Thanks in advance!

sbidoul commented:
Hi @vdewulf my previous comment must be done by someone before merging. It's important to avoid copyright issues with Odoo SA. Can one of your colleagues do it?

vdewulf commented:
Hi @sbidoul. I'll check with my colleagues!

sbidoul committed:
[ADD] setup.py

robinkeunen commented:
Hi @sbidoul 

Is this something like #74 you were expecting?

sbidoul commented:
Closing in favor of #74
